fuse/api.go

```diff
@@ -2,9 +2,74 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// The fuse package provides APIs to implement filesystems in
-// userspace.  Typically, each call of the API happens in its own
+// Package fuse provides APIs to implement filesystems in
+// userspace in terms of raw FUSE protocol.
+//
+// A filesystem is implemented by implementing its server that provides a
+// RawFileSystem interface. Typically the server embeds
+// NewDefaultRawFileSystem() and implements only subset of filesystem methods:
+//
+//	type MyFS struct {
+//		fuse.RawFileSystem
+//		...
+//	}
+//
+//	func MyFSNew() *MyFS {
+//		return &MyFS{
+//			RawFileSystem: fuse.NewDefaultRawFileSystem(),
+//			...
+//		}
+//	}
+//
+//	// Mkdir implements "mkdir" request handler.
+//	//
+//	// For other requests - not explicitly implemented by MyFS - ENOSYS
+//	// will be typically returned to client.
+//	func (fs *MyFS) Mkdir(...) {
+//		...
+//	}
+//
+// Then the filesystem can be mounted and served to a client (typically OS
+// kernel) by creating Server:
+//
+//	fs := MyFSNew() // implements RawFileSystem
+//	fssrv, err := fuse.NewServer(fs, mountpoint, &fuse.MountOptions{...})
+//	if err != nil {
+//		...
+//	}
+//
+//	fssrv.Serve()
+//
+// The server will serve clients by dispatching their requests to the
+// filesystem implementation and conveying responses back. For example "mkdir"
+// FUSE request dispatches to call
+//
+//	fs.Mkdir(*MkdirIn, ..., *EntryOut)
+//
+// "stat" to call
+//
+//	fs.GetAttr(*GetAttrIn, *AttrOut)
+//
+// etc. Please refer to RawFileSystem documentation for details.
+//
+// Typically, each call of the API happens in its own
 // goroutine, so take care to make the file system thread-safe.
+//
+//
+// Higher level interfaces
+//
+// As said above this packages provides way to implement filesystems in terms of
+// raw FUSE protocol. Additionally packages nodefs and pathfs provide ways to
+// implement filesystem at higher levels:
+//
+// Package github.com/hanwen/go-fuse/fuse/nodefs provides way to implement
+// filesystems in terms of inodes. This resembles kernel's idea of what a
+// filesystem looks like.
+//
+// Package github.com/hanwen/go-fuse/fuse/pathfs provides way to implement
+// filesystems in terms of path names. Working with path names is somewhat
+// easier compared to inodes, however do not use pathfs if you care about
+// correctness.
```

Comment:
this needs some more qualification. The tricky bit is what to do with operations that change the name => inode mapping in combination with open files. It can be done correctly, but if using nodefs, you get it for free.